gp_mobo/optimiser.py

```diff
@@ -0,0 +1,149 @@
+import logging
+from typing import Callable, List, Literal, Optional, Union
+
+import numpy as np
+
+from gp_mobo.acquisition_funcs.expected_improvement import (
+    expected_hypervolume_improvement,
+)
+from gp_mobo.acquisition_funcs.hypervolume import Hypervolume, infer_reference_point
+from gp_mobo.acquisition_funcs.pareto import pareto_front
+from gp_mobo.kern_gp.gp_model import independent_tanimoto_gp_predict
+
+ORACLE_LIKE = Callable[[Union[str, List[str]]], Union[float, np.ndarray]]
+ACQUISITION_FUNCTION = Literal["ehvi", "ei"]
+
+
+def ehvi_acquisition(
+    query_smiles: List[str],
+    known_smiles: List[str],
+    known_Y: np.ndarray,
+    gp_means: np.ndarray,
+    gp_amplitudes: np.ndarray,
+    gp_noises: np.ndarray,
+    reference_point: np.ndarray,
+):
+    pred_means, pred_vars = independent_tanimoto_gp_predict(
+        query_smiles=query_smiles,
+        known_smiles=known_smiles,
+        known_Y=known_Y,
+        gp_means=gp_means,
+        gp_amplitudes=gp_amplitudes,
+        gp_noises=gp_noises,
+    )
+
+    pareto_mask = pareto_front(known_Y)
+    pareto_Y = known_Y[pareto_mask]
+
+    ehvi_values = expected_hypervolume_improvement(pred_means, pred_vars, reference_point, pareto_Y)
+
+    return ehvi_values
+
+
+def evaluate_objectives(smiles_list: list[str], oracles: List[ORACLE_LIKE]) -> np.ndarray:
+    """
+    Evaluate the objectives for a list of smiles.
+
+    Given a list of N smiles, return an NxK array A such that
+    A_{ij} is the jth objective function on the ith SMILES.
+
+    """
+    # Initialise arrays for each objective
+    predictions = [np.array(oracle(smiles_list)) for oracle in oracles]
+
+    # Filter out NaN values from f1 and corresponding entries in other arrays
+    f1 = predictions[0]
```

Comment:
So here only the first col could have NaNs?

Reply:
Probably true in her toy examples, I guess we should be more robust!

Reply:
Hilariously (or worryingly) this isn't an issue because if anything has a NaN then we fail the check a couple of lines later that asserts that everything post-cleaning is the same length as f1 pre-cleaning 🤦 . I might have introduced it, or it might just never have been a problem with the datasets in Anabel's project.

Reply:
Nope, was me 🤦